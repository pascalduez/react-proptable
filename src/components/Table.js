// @flow
import React from 'react';
import type { Properties } from '../types';

type Props = {
  properties: Properties,
  className?: string,
};


const Table = ({ properties, className }: Props) => (
  <table className={className}>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Required</th>
        <th>Default</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
      {properties.map(prop => (
        <tr key={prop.name}>
          <td>{prop.name}</td>
          <td>{prop.type.name}</td>
          <td>{String(prop.required)}</td>
          <td>{prop.defaultValue ? prop.defaultValue.value : '-'}</td>
          <td>{prop.description || '-'}</td>
        </tr>
      ))}
    </tbody>
  </table>
);

// const propertiesShape = PropTypes.shape({
//   name: PropTypes.string.isRequired,
//   type: PropTypes.shape({
//     name: PropTypes.string.isRequired,
//   }).isRequired,
//   required: PropTypes.bool,
//   defaultValue: PropTypes.shape({
//     value: PropTypes.string.isRequired,
//   }),
//   description: PropTypes.string,
// });
//
// Table.propTypes = {
//   properties: PropTypes.arrayOf(propertiesShape).isRequired,
//   className: PropTypes.string,
// };

Table.defaultProps = {
  className: '',
};


export default Table;
