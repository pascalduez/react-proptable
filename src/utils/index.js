// @flow
/* eslint-disable no-restricted-syntax, no-continue */
import React from 'react';
import get from 'lodash.get';
import { parse } from 'react-docgen';
import type { ReactComponent, ExtractedProps } from '../types';


export default function getPropTypes(component: ReactComponent): ExtractedProps {
  if (typeof component.__docgenInfo === 'object') {
    return fromDocgen(component);
  }

  if (typeof component === 'string') {
    return fromSource(component);
  }

  return fromReact(component);
}


function fromSource(component: string): ExtractedProps {
  let props = {};

  try {
    props = parse(component).props;
  } catch (ex) {
    // eslint-disable-next-line no-console
    console.error('Woops');
    return props;
  }

  Object.keys(props).forEach((name) => {
    if (!(name in props[name])) {
      props[name].name = name;
    }
  });

  return props;
}


export function fromDocgen(component: ReactComponent): ExtractedProps {
  const props = {};
  const docGenProps = get(component, '__docgenInfo.props');

  for (const [key, val] of Object.entries(docGenProps)) {
    props[key] = val;
    props[key].name = key;
  }

  return props;
}


export function fromReact(component: ReactComponent): ExtractedProps {
  const reactPropTypes = new Map();
  const componentPropTypes = {};

  if (!component.propTypes) {
    return componentPropTypes;
  }

  for (const [name, type] of Object.entries(React.PropTypes)) {
    reactPropTypes.set(type, name);
    // $FlowFixMe
    reactPropTypes.set(type.isRequired, name);
  }

  for (const [name, prop] of Object.entries(component.propTypes)) {
    const type = { name: reactPropTypes.get(prop) || 'custom' };
    // $FlowFixMe
    const required = typeof prop.isRequired !== 'function';
    componentPropTypes[name] = { name, type, required };
  }

  if (!component.defaultProps) {
    return componentPropTypes;
  }

  for (const [name, value] of Object.entries(component.defaultProps)) {
    if (typeof value === 'undefined') {
      continue;
    }

    if (!componentPropTypes[name]) {
      componentPropTypes[name] = { name };
    }

    componentPropTypes[name].defaultValue = { value };
  }

  return componentPropTypes;
}
