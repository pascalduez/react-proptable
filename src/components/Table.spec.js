import React from 'react';
import PropTypes from 'prop-types';
import { render } from 'enzyme';

import { fromReact } from '../utils';
import Table from './Table';

// Fixtures.
// const WithoutProps = () => <span>test</span>;
const WithProps = ({ label, foo }) => <span>{`${label}${foo}`}</span>;
WithProps.displayName = 'WithProps';
WithProps.propTypes = {
  label: PropTypes.string.isRequired,
  foo: PropTypes.number,
};
WithProps.defaultProps = {
  foo: '',
};


describe('<Table>', () => {
  describe('When there is no propType found', () => {
    it('should render the default markup and classes', () => {
      const wrapper = render(
        <Table properties={[]} />
      );

      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('When propTypes are found', () => {
    it('should render the default markup and classes', () => {
      const wrapper = render(
        <Table
          properties={
            Object.values(fromReact(WithProps))
          }
        />
      );

      expect(wrapper).toMatchSnapshot();
    });
  });
});
