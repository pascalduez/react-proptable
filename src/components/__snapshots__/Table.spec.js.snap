// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table> When propTypes are found should render the default markup and classes 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Type
      </th>
      <th>
        Required
      </th>
      <th>
        Default
      </th>
      <th>
        Description
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        label
      </td>
      <td>
        <pre>
          <code>
            string
          </code>
        </pre>
      </td>
      <td>
        yes
      </td>
      <td>
        <pre>
          <code>
            -
          </code>
        </pre>
      </td>
      <td>
        -
      </td>
    </tr>
    <tr>
      <td>
        foo
      </td>
      <td>
        <pre>
          <code>
            number
          </code>
        </pre>
      </td>
      <td>
        no
      </td>
      <td>
        <pre>
          <code />
        </pre>
      </td>
      <td>
        -
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`<Table> When there is no propType found should render the default markup and classes 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Type
      </th>
      <th>
        Required
      </th>
      <th>
        Default
      </th>
      <th>
        Description
      </th>
    </tr>
  </thead>
  <tbody />
</table>
`;
